Code/Geometry/Trajectory.cpp

```diff
@@ -0,0 +1,195 @@
+//
+// Copyright (C) 2003-2016 Paolo Tosco
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <RDGeneral/Invariant.h>
+#include <RDGeneral/BadFileException.h>
+#include <fstream>
+#include <sstream>
+#include <set>
+#include "Trajectory.h"
+
+namespace RDGeom {
+
+Snapshot::Snapshot(boost::shared_array<double> pos, double energy) :
+  d_trajectory(NULL),
+  d_energy(energy),
+  d_pos(pos) {}
+
+Point2D Snapshot::getPoint2D(unsigned int pointNum) const {
+  PRECONDITION(d_pos, "d_pos must not be NULL");
+  PRECONDITION(d_trajectory, "d_trajectory must not be NULL");
+  PRECONDITION(d_trajectory->dimension() == 2, "d_dimension must be == 2");
+  PRECONDITION(d_trajectory->numPoints(), "d_numPoints must be > 0");
+  URANGE_CHECK(pointNum, d_trajectory->numPoints() - 1);
+  unsigned int i = pointNum * d_trajectory->dimension();
+  return Point2D(d_pos[i], d_pos[i + 1]);
+}
+
+Point3D Snapshot::getPoint3D(unsigned int pointNum) const {
+  PRECONDITION(d_pos, "d_pos must not be NULL");
+  PRECONDITION(d_trajectory, "d_trajectory must not be NULL");
+  PRECONDITION(d_trajectory->dimension() >= 2, "d_dimension must be >= 2");
+  PRECONDITION(d_trajectory->numPoints(), "d_numPoints must be > 0");
+  URANGE_CHECK(pointNum, d_trajectory->numPoints() - 1);
+  unsigned int i = pointNum * d_trajectory->dimension();
+  return (Point3D(d_pos[i], d_pos[i + 1],
+          (d_trajectory->dimension() == 3) ? d_pos[i + 2] : 0.0));
+}
+
+Trajectory::Trajectory(unsigned int dimension, unsigned int numPoints) :
+  d_dimension(dimension),
+  d_numPoints(numPoints) {}
+
+Trajectory::Trajectory(const Trajectory &other) :
+  d_dimension(other.d_dimension),
+  d_numPoints(other.d_numPoints) {
+  for (SnapshotSPtrVect::const_iterator vectIt = other.d_snapshotVect.begin();
+    vectIt != other.d_snapshotVect.end(); ++vectIt)
+    addSnapshot(new Snapshot(*(*vectIt)));
+}
+
+unsigned int Trajectory::addSnapshot(Snapshot *s) {
+  return insertSnapshot(d_snapshotVect.size(), s);
+}
+
+Snapshot *Trajectory::getSnapshot(unsigned int snapshotNum) const {
+  URANGE_CHECK(snapshotNum, d_snapshotVect.size() - 1);
+  return d_snapshotVect[snapshotNum].get();
+}
+
+unsigned int Trajectory::insertSnapshot(unsigned int snapshotNum, Snapshot *s) {
+  URANGE_CHECK(snapshotNum, d_snapshotVect.size());
+  s->d_trajectory = this;
```

Comment:
I had this concern too, so I thought to see what happened with Conformer and ROMol object. Apparently a Conformer cannot be owned by multiple molecules, unless you make a copy. If you uncomment the commented line in the snippet below, it will segfault:

```
/*
Can be built with:
g++ -Wall -Wno-unused-function -o testConformer testConformer.cpp \
  -I$RDBASE/Code -I. -L$RDBASE/lib -lFileParsers -lSmilesParse \
  -lRDGeneral -lGraphMol -lboost_system -lboost_thread
*/


#include <GraphMol/RDKitBase.h>
#include <GraphMol/FileParsers/MolSupplier.h>
#include <iostream>


using namespace RDKit;

int main(int argc, char *argv[])
{
  if (argc < 2) {
    std::cerr << "Usage: testConformer <molecule.sdf>" << std::endl;
    return -1;
  }
  std::string sdf(argv[1]);
  SDMolSupplier supplier(sdf, true, false);
  for (unsigned int i = 0; i < supplier.length(); ++i) {
    ROMol *mol = supplier[i];
    Conformer *conf = new Conformer(mol->getConformer());
    ROMol *mol2 = new ROMol(*mol);
    mol->addConformer(conf);
    Conformer *conf2 = new Conformer(*conf);
    mol2->addConformer(conf2);
    //mol2->addConformer(conf);
    delete mol2;
    delete mol;
  }
  
  return 0;
}

```
So I sticked to the same paradigm: a Snapshot cannot be in multiple trajectories. Happy to modify this based on Greg's and Brian's inputs.